scipy/sparse/linalg/_dsolve/_superlumodule.c

```diff
@@ -263,6 +264,173 @@ static PyObject *Py_gstrf(PyObject * self, PyObject * args,
+        return NULL;
+    }
+
+    int L_type = PyArray_TYPE((PyArrayObject*)L_nzvals);
+    int U_type = PyArray_TYPE((PyArrayObject*)U_nzvals);
+    if (L_type != U_type) {
+        PyErr_SetString(PyExc_TypeError,
+                        "nzvals types of L and U differ");
+        return NULL;
+    }
+    if (!CHECK_SLU_TYPE(L_type)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "nzvals is not of a type supported by SuperLU");
+        return NULL;
+    }
+
+    /* Create SuperLU matrices out of L and U. */
+    int* L_col_to_sup = intMalloc(L_N+1), *L_sup_to_col = intMalloc(L_N+1);
+    for(int i=0; i<=L_N; i++){
+        L_col_to_sup[i] = i;
+        L_sup_to_col[i] = i;
+    }
+    L_col_to_sup[L_N] = L_N-1;
+    SuperMatrix L_super = {0}, U_super={0};
+    int L_conv_err = SparseFormat_from_spMatrix(&L_super, L_N, L_N, L_nnz, -1,
+        (PyArrayObject*)L_nzvals, (PyArrayObject*)L_rowind, (PyArrayObject*)L_colptr,
+        L_type, SLU_SC, SLU_TRLU, L_col_to_sup,L_sup_to_col);
+    if (L_conv_err) {
+        return NULL;
+    }
+    int U_conv_err = SparseFormat_from_spMatrix(&U_super, U_N, U_N, U_nnz, 0,
+        (PyArrayObject*)U_nzvals, (PyArrayObject*)U_rowind, (PyArrayObject*)U_colptr,
+        U_type, SLU_NC, SLU_TRU, NULL,NULL);
+    if (U_conv_err) {
+        Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+        return NULL;
+    }
+
+    /* Read right-hand-side/solution vector. */
+    PyArrayObject* X_arr = (PyArrayObject*)PyArray_FROMANY(
+        (PyObject*)X_py, L_type, 1, 2,
+        NPY_ARRAY_F_CONTIGUOUS | NPY_ARRAY_ENSURECOPY);
+    if (X_arr == NULL) {
+        SUPERLU_FREE((void*)L_col_to_sup);
+        SUPERLU_FREE((void*)L_sup_to_col);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+        return NULL;
+    }
+    if (PyArray_DIM((PyArrayObject*)X_arr, 0) != L_N) {
+        PyErr_SetString(PyExc_ValueError,
+                        "right hand side array has invalid shape");
+        SUPERLU_FREE((void*)L_col_to_sup); SUPERLU_FREE((void*)L_sup_to_col);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+        Py_DECREF(X_arr);
+        return NULL;
+    }
+
+    SuperMatrix X;
+    if (DenseSuper_from_Numeric((SuperMatrix*)&X, (PyObject*)X_arr)) {
+        SUPERLU_FREE((void*)L_col_to_sup); SUPERLU_FREE((void*)L_sup_to_col);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+        Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+        Py_DECREF(X_arr);
+        return NULL;
+    } /* X and X_arr share the same data but X_arr "owns" it. */
+
+    /* Call SuperLU functions. */
+    int info=0;
+    SuperLUStat_t stat = { 0 };
+    StatInit((SuperLUStat_t *)&stat);
+    int* perm_c = intMalloc(L_N);
+    for (int i=0; i<L_N; i++) {
+        perm_c[i] = i;
+    }
+    int* perm_r = perm_c;
```

Comment:
Right, the SuperLU documentation is silent about whether the row and column permutation can be the same vector, and the signature declares them as `int*` not `int const*`. However, the implementation in [dgstrs.c](https://github.com/scipy/scipy/blob/ed65b675496cb38bf021858d56e32a7d740ca2ea/scipy/sparse/linalg/_dsolve/SuperLU/SRC/dgstrs.c) (and variants for other types) only read `perm_c` and `perm_r` and do not write to it. That's why it still works. I've opened an issue upstream that `perm_c` and `perm_r` can be const pointers (xiaoyeli/superlu#142).

Would you suggest to nevertheless be safe and allocate and initialize two separate copies of the array?

Reply:
Great, thanks for checking and opening that issue. No, I think we're safe here, so let's leave things as is since this is easier and more performant. In case the SuperLU author says it's not future-proof, we can update later.